Adopt a new way to go with the loglevels

DevUt opened the pull request.

DevUt committed:
Move <unistd.h> above other headers as it defines POSIX_VERSION which other headers check

DevUt committed:
Fix pointer bug in strcpy and change the way of concentating the tag and message

The new way of concenatating is more sane and safe work

In the new way we allocate the string literal a space then strcpy the tag into it and strcat the the true message with the tag

In the end when klog finishes and returns a value(int) it would first free the buffer to avoid memeory cluttering and then return the value that klog returned in the first place

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -13,14 +13,14 @@
   if (outbuf > 0) // Trying to open KMSG,
   {
-    strcpy(buf,*msg);
+    strcpy(buf,msg);
```
btw we still didn't check msg != NULL

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
+*/
+int logwarn(char *warning)
+{
+ char *warn = malloc(256);
```
Still bad indent :s

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
+  strcat(warn ,warning);
+ int ret = klog(warn) ;
+  return(ret);
+  free(warn);
```
Unreachable code

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -13,14 +13,14 @@
+ char *buf = malloc(256);
+ int ret = -1;
 
   // Write to kmsg if possibe.
```
You didn't remove comments anymore?

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -51,32 +51,59 @@ int klog(char *msg)
+  */
+int logdebug (char *debug_msg)
+{
+ char *debug = malloc(256);
```
Your are using malloc everywhere, good catch. 
But you set the same value on all entries. Can you use a preprocessor variable instead please?

nailyk-fr reviewed (changes requested):
Really hard to review. Patch is unreadable because of the indent problem. 
Your new version looks good except the indent.

DevUt committed:
Remove loglevel functions and add loglevel variable && remove comments

Now we no longer use functions for loglevel there are loglevel variable that you pass to the klog function

We removed the comments ; One a single breifing about the funtions doing above the function is a whole lot good

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -8,102 +8,64 @@
-    */
-    if (access(LOG_LOC, F_OK ) == -1) {  // Check if the file exists.
-      outbuf = open(LOG_LOC, O_CREAT | O_WRONLY); // If it doesn't exist, create it and open it to write to it.
+    switch (level)
```
Switch misplaced, need to be out of the if(open())else statement.

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -8,102 +8,64 @@
+      case 3 : tag = "Debug :"
+                strcat(tag,msg);
+              break ;
+      default : tag = "Error : Unknown log level "
```
Please two point: 
Make every log level same string lenght. (like ERR, WAR, INF)
I think you can import kernel log level too and assume default as max: 

case 7: 
default: 
    foo bar; 
    break;

DevUt replied:
I didn't get the default point please elaborate

nailyk-fr replied:
Don't mind, will push them afterward.

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -8,102 +8,64 @@
-
-	close(outbuf); // CLOSE THE FILE DISCRIPTOR TO AVOID LEAK ;
-
-	free(buf); // Empty buf memory.
```
no free(tag)?

DevUt replied:
Why free ? 
I used stack memory instead of heap 
I have no control on it (and its good) 
Its premanaged

nailyk-fr replied:
Forget it. Yes and this way is way better!

nailyk-fr reviewed (commented):
This one looks good, except the misplaced switch statement. 
Please, only for my happiness, can you rename tag --> buf?

DevUt committed:
Rename buf to tag

DevUt committed:
Add option to Wrap log tags and fix assignment of buf

After some thought it was decided that the log tags should all be of the same length
> but for some it would become too complex of understanding I ( considering a noob user reading the log)

So it was decided to add a flag that can change the way it behaves

 So we added the LOG_WRAP_TAGS flag ,so that the use can the dev can enable it in the makefiles and have the shortened tags with Info as I: Warn as W: Error as E: and Debug as D:

We also fixed assignment of buf , it was getting assigned to a string literal while being a buffer, so we utilized strcpy over here to assign it

DevUt committed:
Move the Switch the out of the if-else statement

DevUt committed:
Add header DEAFULT_LOG_LOC and change LOG_LOc to ALT_LOG_LOC

Previously if the user wanted open anything else other than /dev/kmsg , he couldnt. He had to edit the source for changing the fle to be opned. Now he could just add DEFAULT_LOG_LOC to his device specific makefile to oprn a particular location to write. In the same way he can also now define ATL_LOG_lOC to open a file if /dev/kmsg fails

By default DEFAULT_LOG_LOC is /dev/kmsg and ALT_LOG_LOC is /cache/lible.log

DevUt committed:
Commentify

nailyk-fr commented on C/kmsg_log.c:
```diff
@@ -19,28 +24,28 @@
  char buf [256];
 
- int outbuf = open("/dev/kmsg", O_WRONLY);
+ int outbuf = open(DEFAULT_LOG_LOC, O_WRONLY);
```
Force define of DEFAULT_LOG_LOC in case not passed to makefile.

nailyk-fr commented on C/Android.mk:
```diff
@@ -2,4 +2,8 @@ LOCAL_PATH := $(call my-dir)
 endif
+
+ifeq ($(DEFAULT_LOG_LOC),)
+	LOCAL_CFLAGS += -DDEFAULT_LOG_LOC=\"$(DEFAULT_LOG_LOC)\"
```
Add possibility to define ALT_LOG_LOC too.

DevUt committed:
Add ALT_LOG_LOC in the Android.mk
